storage2/src/snapshot.rs

```diff
@@ -8,58 +14,121 @@ use crate::state::StateRead;
-        Self {
+        Self(Inner {
             rocksdb_snapshot,
             jmt_version,
             db,
-        }
+        })
     }
 
     pub fn jmt_version(&self) -> jmt::Version {
-        self.jmt_version
+        self.0.jmt_version
     }
 }
 
+#[async_trait]
 impl StateRead for Snapshot {
     /// Fetch a key from the JMT column family.
-    fn get_raw(&self, key: &str) -> Result<Option<Vec<u8>>> {
-        let jmt_cf = self
-            .db
-            .cf_handle("jmt")
-            .expect("jmt column family not found");
-        self.rocksdb_snapshot
-            .get_cf(jmt_cf, key)
-            .map_err(Into::into)
+    async fn get_raw(&self, key: &str) -> Result<Option<Vec<u8>>> {
+        let span = Span::current();
+        let db = self.0.db;
+        let rocksdb_snapshot = self.0.rocksdb_snapshot.clone();
+        let key = key.to_string();
+        tokio::task::Builder::new()
+            .name("Snapshot::get_raw")
+            .spawn_blocking(move || {
+                span.in_scope(|| {
+                    let jmt_cf = db.cf_handle("jmt").expect("jmt column family not found");
+                    rocksdb_snapshot.get_cf(jmt_cf, key).map_err(Into::into)
+                })
+            })?
+            .await?
     }
 
-    /// Fetch a key from the sidecar column family.
-    fn get_sidecar(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
-        let sidecar_cf = self
-            .db
-            .cf_handle("sidecar")
-            .expect("sidecar column family not found");
-        self.rocksdb_snapshot
-            .get_cf(sidecar_cf, key)
-            .map_err(Into::into)
+    /// Fetch a key from the nonconsensus column family.
+    async fn get_nonconsensus(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        let span = Span::current();
+        let db = self.0.db;
+        let rocksdb_snapshot = self.0.rocksdb_snapshot.clone();
+        let key: Vec<u8> = key.to_vec();
+        tokio::task::Builder::new()
+            .name("Snapshot::get_nonconsensus")
+            .spawn_blocking(move || {
+                span.in_scope(|| {
+                    let nonconsensus_cf = db
+                        .cf_handle("nonconsensus")
+                        .expect("nonconsensus column family not found");
+                    rocksdb_snapshot
+                        .get_cf(nonconsensus_cf, key)
+                        .map_err(Into::into)
+                })
+            })?
+            .await?
+    }
+
+    async fn prefix_raw(
+        &self,
+        prefix: &str,
+    ) -> Result<Pin<Box<dyn Stream<Item = (String, Box<[u8]>)> + Send + '_>>> {
+        let span = Span::current();
+        let db = self.0.db;
+        let rocksdb_snapshot = self.0.rocksdb_snapshot.clone();
+        let mut options = rocksdb::ReadOptions::default();
+        options.set_iterate_range(rocksdb::PrefixRange(prefix.as_bytes()));
+        let mode = rocksdb::IteratorMode::Start;
+
+        let (tx, rx) = mpsc::channel(100);
```

Comment:
I wonder what the appropriate value to set here is. Should this be configurable, so we can tune and benchmark it?

Reply:
Configurable in what sense, as a command-line flag?

Reply:
I don't think we should necessarily have a command line flag, but it might be good to (as a maintenance task) ferret out such "magic constants" and perhaps have a configuration file that defines them, with some sensible defaults. I have no idea if 100 is the best buffer size here, and similarly for other places in the codebase where we arbitrarily pick one; it might be worth experimentally finding out.

Reply:
10 is a good number, it's good to avoid buffer bloat. I don't think we'll want a configuration file, we can just pick something now and do performance tuning later/never.